### PR TITLE
Exceptions

### PR DESCRIPTION
**Pull request by dbhart**
## Summary
Provides updated EpanetException class and subclasses that allow for cleaner error reporting during IO.

### CODE REVIEW

**Comment by dbhart** [comment]
@kbonney I see you also have a PR in to do some work on Exceptions. I know that the request with MSX was that we somehow use the same error codes as EPANET does in wntr.epanet.msx.io, and I started setting that up for wntr.epanet.io also, but I think we did this in parallel and should try to merge our approaches. :-)

**Comment by kbonney** [comment]
Thanks for bringing this to my attention @dbhart. I think we can stick with what you've proposed here and I can adjust the contents of #374 to fit within that framework.

**Comment by dbhart** [comment]
Trim down (remove Enum) and create some tests

File: wntr/epanet/toolkit.py [hunk 1]
```diff
@@ -538,21 +513,6 @@ def ENgetflowunits(self):
...
-        :return the node's ID name.
-        """
-        fValue = ctypes.create_string_buffer(SizeLimits.EN_MAX_ID.value)
-        if self._project is not None:
-            self.errcode = self.ENlib.EN_getnodeid(self._project, iIndex, byref(fValue))
-        else:
-            self.errcode = self.ENlib.ENgetnodeid(iIndex, byref(fValue))
-        self._error()
-        return str(fValue.value, 'UTF-8')
-
```
Comment (kaklise) [inline wntr/epanet/toolkit.py:555]: why was this removed?

File: wntr/epanet/toolkit.py [hunk 2]
```diff
@@ -48,41 +46,15 @@ def ENgetwarning(code, sec=-1):
...
-                + "Vavles cannot deliver enough flow - one or more flow control valves could not deliver the required flow even when fully open"
-        )
-    elif code == 6:
-        return (
-                header
-                + "System has negative pressures - negative pressures occurred at one or more junctions with positive demand"
-        )
+        header = "{}".format(code)
+    if code < 100:
+        msg = EN_ERROR_CODES.get(code, 'Unknown warning %s')
```
Comment (kaklise) [inline wntr/epanet/toolkit.py:53]: Why are these unknown warnings? Error codes 1-6 are listed as simulation warnings.

File: wntr/epanet/toolkit.py [hunk 3]
```diff
@@ -574,22 +534,6 @@ def ENgetnodeindex(self, sId):
...
-        :return int node type
-        """
-        fValue = ctypes.c_int()
-        if self._project is not None:
-            self.errcode = self.ENlib.EN_getnodetype(self._project, iIndex, byref(fValue))
-        else:
-            self.errcode = self.ENlib.ENgetnodetype(iIndex, byref(fValue))
-        self._error()
-        return fValue.value
-
```
Comment (kaklise) [inline wntr/epanet/toolkit.py:592]: why was this removed?

File: wntr/epanet/toolkit.py [hunk 4]
```diff
@@ -636,21 +580,6 @@ def ENgetlinkindex(self, sId):
...
-        :return:linkType
-        """
-        fValue = ctypes.c_int()
-        if self._project is not None:
-            self.errcode = self.ENlib.EN_getlinktype(self._project, iIndex, byref(fValue))
-        else:
-            self.errcode = self.ENlib.EN_getlinktype(iIndex, byref(fValue))
-        self._error()
-        return fValue.value
-
```
Comment (kaklise) [inline wntr/epanet/toolkit.py:653]: why was this removed?

File: wntr/epanet/toolkit.py [hunk 5]
```diff
@@ -768,6 +697,173 @@ def ENgettimeparam(self, eParam):
...
+        iType : int
+            _description_
+        iLinkIndex : int
+            _description_
+        dSetting : float
+            _description_
+        iNodeIndex : int
+            Set to 0 for time of day or timer
+        dLevel : float
+            _description_
```
Comment (kaklise) [inline wntr/epanet/toolkit.py:798]: update descriptions

File: wntr/epanet/toolkit.py [hunk 6]
```diff
@@ -768,6 +697,173 @@ def ENgettimeparam(self, eParam):
...
+                ctypes.c_int(iLinkIndex), 
+                ctypes.c_double(dSetting), 
+                ctypes.c_int(iNodeIndex), 
+                ctypes.c_double(dLevel)
+            )
+        self._error()
+
+    def ENdeletecontrol(self, iControlIndex):
+        """
+        Get a time parameter value
```
Comment (kaklise) [inline wntr/epanet/toolkit.py:842]: Why is this a time value?

File: wntr/epanet/toolkit.py [hunk 7]
```diff
@@ -768,6 +697,173 @@ def ENgettimeparam(self, eParam):
...
+        iType : int
+            _description_
+        iLinkIndex : int
+            _description_
+        dSetting : float
+            _description_
+        iNodeIndex : int
+            Set to 0 for time of day or timer
+        dLevel : float
+            _description_
```
Comment (kaklise) [inline wntr/epanet/toolkit.py:715]: update descriptions and add a test for add/get/set controls using the toolkit.

File: wntr/epanet/exceptions.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+            The value that is invalid
+        args : additional non-keyword arguments, optional
+            If there is a string-format within the error code's text, these will be used to 
+            replace the format, otherwise they will be output at the end of the Exception message.
+        line_num : int, optional
+            The line number, if reading an INP file, by default None
+        line : str, optional
+            The contents of the line, by default None
+        """
+        super().__init__(code, value, *args, line_num=line_num, line=line)
```
Comment (kaklise) [inline wntr/epanet/exceptions.py:301]: add a test that covers some exceptions

File: wntr/epanet/toolkit.py (hunk 2 above)
Comment (dbhart) [reply wntr/epanet/toolkit.py:53]: These are not unknown warnings, the ``get`` argument is passing a default, which means if someone passes 17, for example, we don't know what warning 17 is. If it is 1-6, then it will give the actual warning, not just the "unknown warning" piece.

File: wntr/epanet/toolkit.py (hunk 5 above)
Comment (dbhart) [reply wntr/epanet/toolkit.py:798]: Will do

File: wntr/epanet/toolkit.py (hunk 7 above)
Comment (dbhart) [reply wntr/epanet/toolkit.py:715]: Will do

File: wntr/epanet/toolkit.py (hunk 4 above)
Comment (dbhart) [reply wntr/epanet/toolkit.py:653]: Should not have been, will get fixed.

File: wntr/epanet/toolkit.py (hunk 3 above)
Comment (dbhart) [reply wntr/epanet/toolkit.py:592]: Should not have been.

File: wntr/epanet/toolkit.py (hunk 1 above)
Comment (dbhart) [reply wntr/epanet/toolkit.py:555]: I don't know how this got removed, and I'm trying to fix that.

File: wntr/tests/test_epanet_exceptions.py
Comment (dbhart) [inline wntr/tests/test_epanet_exceptions.py:1]: @kaklise added tests for exceptions

**Comment by dbhart** [comment]
Added documentation file, links to it, tests, and the files are now all up to date.

**Comment by dbhart** [comment]
Todo to finish this PR: 
* Get rid of errors.rst file 
* Finish testing
